Parallel Apply : Updating content for 5.2 (#4458)
EnterpriseDB/docs: product_docs/docs/pgd/5/parallelapply.mdx

---
title: Parallel apply
navTitle: Parallel apply
---

### What is Parallel apply?

!!! Note Improved in PGD 5.2

> This feature has been enhanced in PGD 5.2 (and later) and is recommended for more scenarios, especially with commit scopes. 

Reviewer: I'd nix the "(and later)"

Author: Now nixed

!!!

Parallel apply is a feature of PGD that allows a PGD node to use multiple writers per subscription. This generally increases the throughput of a subscription and improves replication performance.

The transactional changes from the subscription are written by the multiple Parallel apply writers. However, each writer ensures that the final commit of its transaction does not violate the commit order as executed on the origin node. If there is a violation, an error is generated and the transaction can be rolled back. 

This mechanism has previously meant that if a transaction is pending commit and modifies a row that another transaction needs to change, and that other transaction executed on the origin node before the pending transaction did, then the resulting error could manifest itself as a deadlock if the pending transaction has taken out a lock request. 

> Additionally, handling the error could increase replication lag, due to a combination of the time taken to detect the deadlock, for the client to rollback its transaction, indirect garbage collection of the already applied changes and time taken to redo the work. 

Reviewer: ```suggestion
Additionally, handling the error could increase replication lag, due to a combination of the time taken to detect the deadlock, for the client to rollback its transaction, indirect garbage collection of the already applied changes, and time taken to redo the work. 
```

This is where Parallel apply’s deadlock mitigation, introduced in PGD 5.2, comes into play.

Parallel apply’s deadlock mitigation is always enabled. Parallel apply deadlock mitigation gets transactions to preview their write sets before they are executed. Transactions can then probe other contemporary transactions with older commit timestamps looking for a write set intersection. 

> Where there are intersections, Parallel apply will execute the transaction until it finds a conflicting row. Only then will it put the transaction on hold until the intersecting transactions have committed the conflicted rows. It will then continue applying row changes from the transaction.

Reviewer: ```suggestion
Where there are intersections, Parallel apply executes the transaction until it finds a conflicting row. Only then will it put the transaction on hold until the intersecting transactions have committed the conflicted rows. It then continues applying row changes from the transaction.
```

### What is Parallel Apply?

!!! Note Improved in PGD 5.2
This feature has been enhanced in PGD 5.2 and is recommended for more scenarios. 
!!!

Parallel Apply is a feature of PGD that allows a PGD node to use multiple writers per subscription. This behavior generally increases the throughput of a subscription and improves replication performance.

> The transactional changes from the subscription are written by the multiple Parallel Apply writers. However, each writer ensures that the final commit of its transaction doesn't violate the commit order as executed on the origin node. If there's a violation, an error is generated and the transaction can be rolled back. 

Reviewer 2: are they "multiple parallel apply writers" or when there are multiple writers we call it Parallel Apply - meaning if the the number of writers is 1 then we're not using paralell apply but any scenario where it's greater than 1 we're applying writes in parallel which we refer to as Parallel Apply - and when there's more than one writer there's these extra things we take into account

Author: ```suggestion
When Parallel Apply is operating, the transactional changes from the subscription are written by multiple writers. However, each writer ensures that the final commit of its transaction doesn't violate the commit order as executed on the origin node. If there's a violation, an error is generated and the transaction can be rolled back. 
```

This mechanism has previously meant that if a transaction is pending commit and modifies a row that another transaction needs to change, and that other transaction executed on the origin node before the pending transaction did, then the resulting error could manifest itself as a deadlock if the pending transaction has taken out a lock request. 

Additionally, handling the error could increase replication lag, due to a combination of the time taken to detect the deadlock, the time taken for the client to rollback its transaction, the time taken for indirect garbage collection of the already applied changes, and the time taken to redo the work. 

This is where Parallel Apply’s deadlock mitigation, introduced in PGD 5.2, comes into play.

> Parallel Apply’s deadlock mitigation is always enabled. Parallel Apply deadlock mitigation gets transactions to preview their write sets before they are executed. Transactions can then probe other contemporary transactions with older commit timestamps looking for a write set intersection. 

Author: ```suggestion
For any transaction, Parallel Apply now looks at upcoming writes, and where there is a potential conflict, places the transaction on hold until the conflict has passed.
```

> Where there are intersections, Parallel Apply executes the transaction until it finds a conflicting row. Only then will it put the transaction on hold until the intersecting transactions have committed the conflicted rows. It then continues applying row changes from the transaction.

Author: ```suggestion
```

> 

Reviewer 3: I agree with PM's input this morning regarding implementation details not being part of external doc. IMHO, it's not so much that IP or trade secrets might be revealed to competitors as it is keeping it simple to describe the value proposition how to put the feature to use.

OTOH, PA deadlock mitigation is something to highlight in the 5.2 release notes. Many users may have put the feature aside because it did not work for them; they should be encouraged to give it another try.

Author: Taking content out at this point is problematic because there's no concept documentation to fall back to. And this is a detail, which as you point out fixes an existing issue which may have put users aside.

The current release notes format isn't built for anything more than a cursory highlight and I would suggest, at least for 5.2 that we leave this in place till we have concept documentation that can allow us the cut down on the implementation detail.

Author: I sat down and reduced the detail in the deadlock mitigation to a simple "It sees a conflict, it puts it on hold." - This should suffice for the user manual.

SELECT bdr.alter_subscription_enable ('bdr_bdrdb_bdrgroup_node2_node1');
```

### When to use Parallel apply

Parallel apply is always on by default and, for most operations, we recommend that it's left on.

> From PGD 5.2, Parallel apply works with Group Commit, CAMO, and eager replication. 

Reviewer 4: Sadly, we had to postpone 2/3 of this. Please replace with something along the lines of:

```
From PGD 5.2, Parallel apply works with CAMO. It is not compatible with
Group Commit nor eager replication. It must be disabled if Group Commit
or eager replication is used.
```

Author: Fixed

SELECT bdr.alter_subscription_enable ('bdr_bdrdb_bdrgroup_node2_node1');
```

### When to use Parallel apply

> Parallel apply is always on by default and for most operations, we recommend that it is left on.

Reviewer: ```suggestion
Parallel apply is always on by default and, for most operations, we recommend that it's left on.
```

### When not to use Parallel apply

> From PGD 5.2, Parallel apply works with Group Commit, CAMO and eager replication. 

Reviewer: ```suggestion
From PGD 5.2, Parallel apply works with Group Commit, CAMO, and eager replication. 
```
Seems odd to capitalize Group Commit and not eager replication.  Again, I would think you would capitalize both or neither.

How does this text tell you when **not** to use Parallel apply?

Author: Section exists in the previous version of the page already in place. It's also there in the expectation that at some point soon, we are going to find circumstances when PA isn't recommended. As this is still a draft, I've been holding it open.

### Monitoring Parallel apply

> To support Parallel apply's deadlock mitigation, PGD 5.2 adds columns to [`bdr.stat_subscription`](/pgd/latest/reference/catalogs-visible#bdrstat_subscription). The new columns are `nprovisional_waits`, `ntuple_waits` and `ncommmit_waits`. The `nprovisional_waits` value reflects the number of operations on the same tuples being performed by concurrent apply transactions. These are provisional waits which are not actually waiting yet but could. 

Reviewer: ```suggestion
To support Parallel apply's deadlock mitigation, PGD 5.2 adds columns to [`bdr.stat_subscription`](/pgd/latest/reference/catalogs-visible#bdrstat_subscription). The new columns are `nprovisional_waits`, `ntuple_waits` and `ncommmit_waits`. The `nprovisional_waits` value reflects the number of operations on the same tuples being performed by concurrent apply transactions. These are provisional waits which aren't actually waiting yet but could. 
```

> If apply operations are affected by the deadlock mitigation measures, they will wait until they can be safely applied and are counted in `ntuple_waits`. Fully applied transactions which waited before being committed are counted in `ncommit_waits`. 

Reviewer: ```suggestion
If apply operations are affected by the deadlock mitigation measures, they wait until they can be safely applied and are counted in `ntuple_waits`. Fully applied transactions which waited before being committed are counted in `ncommit_waits`. 
```

Author: I think the "will" is appropriate as, as I understand it they won't be waiting quite yet (there's some prediction going on) but they will wait when then time comes for them to be applied.

### When to use Parallel Apply

Parallel Apply is always on by default and, for most operations, we recommend that it's left on.

> From PGD 5.2, Parallel Apply works with CAMO. It is not compatible with Group Commit or eager replication and should be disabled if Group Commit or eager replication are in use. 

Reviewer 2: we don't say which commits scopes/replication types paralell apply can be used for before this - i believe it's asynch, legacy synch (which may or may not include sync commit) and now group commit)

Author: @cwcaro - do you know what things PA worked with correctly in the 5.0/5.1 series?

Reviewer 3: Not sure, but I do know that PA also worked with streaming of large transactions

### Monitoring Parallel Apply

To support Parallel Apply's deadlock mitigation, PGD 5.2 adds columns to [`bdr.stat_subscription`](/pgd/latest/reference/catalogs-visible#bdrstat_subscription). The new columns are `nprovisional_waits`, `ntuple_waits` and `ncommmit_waits`. The `nprovisional_waits` value reflects the number of operations on the same tuples being performed by concurrent apply transactions. These are provisional waits which aren't actually waiting yet but could. 

> If apply operations are affected by the deadlock mitigation measures, they will wait until they can be safely applied and are counted in `ntuple_waits`. Fully applied transactions which waited before being committed are counted in `ncommit_waits`. 

Reviewer 2: shouldn't there still be a warning about not using it with Group Commit and Eager and CAMO before 5.1?

Author: Fixed further down in the text.
